README.md

# What is Neurosynth?

Neurosynth is a Python package for large-scale synthesis of functional neuroimaging data.

## Code status

* [![tests status](https://secure.travis-ci.org/neurosynth/neurosynth.png?branch=master)](https://travis-ci.org/neurosynth/neurosynth) travis-ci.org (master branch)

* [![Coverage Status](https://coveralls.io/repos/neurosynth/neurosynth/badge.png?branch=master)](https://coveralls.io/r/neurosynth/neurosynth?branch=master)

## Installation

Dependencies:

* NumPy/SciPy
* pandas
* NiBabel
* [ply](http://www.dabeaz.com/ply/)
* scikit-learn

We recommend installing the core scientific packages (NumPy, SciPy, pandas, sklearn) via a distribution like [https://store.continuum.io/cshop/anaconda/](Anaconda), which will keep clutter and conflicts to a minimum. The other packages can be installed with pip using the requirements file:

	> pip install -r requirements.txt

Or by name:

	> pip install nibabel ply

Assuming you have those packages in working order, the easiest way to install Neurosynth is from the command line with pip:

	> pip install neurosynth

Alternatively, if you want the latest development version, you can install directly from the github repo:

	> pip install git+https://github.com/neurosynth/neurosynth.git

Depending on your operating system, you may need superuser privileges (prefix the above line with 'sudo').

That's it! You should now be ready to roll.

## Documentation

Documentation, including a [full API reference](http://neurosynth.readthedocs.org/en/latest/reference.html), is available [here](http://neurosynth.readthedocs.org/en/latest/)(caution: work in progress).

## Usage

Running analyses in Neurosynth is pretty straightforward. We're working on a user manual; in the meantime, you can take a look at the code in the /examples directory for an illustration of some common uses cases (some of the examples are in IPython Notebook format; you can view these online by entering the URL of the raw example on github into the online [IPython Notebook Viewer](http://nbviewer.ipython.org)--for example [this tutorial](http://nbviewer.ipython.org/urls/raw.github.com/neurosynth/neurosynth/master/examples/neurosynth_demo.ipynb) provides a nice overview). The rest of this Quickstart guide just covers the bare minimum.

The NeuroSynth dataset resides in a separate submodule located [here](http://github.com/neurosynth/neurosynth-data). Probably the easiest way to get the most recent data though is from within the Neurosynth package itself:

	import neurosynth as ns
	ns.dataset.download(path='.', unpack=True)


...which should download the latest database files and save them to the current directory. Alternatively, you can manually download the data files from the [neurosynth-data repository](http://github.com/neurosynth/neurosynth-data). The latest dataset is always stored in current_data.tar.gz in the root folder. Older datasets are also available in the archive folder.

The dataset archive (current_data.tar.gz) contains 2 files: database.txt and features.txt. These contain the activations and meta-analysis tags for Neurosynth, respectively.

Once you have the data in place, you can generate a new Dataset instance from the database.txt file:

	> from neurosynth.base.dataset import Dataset
	> dataset = Dataset('data/database.txt')

This should take several minutes to process. Note that this is a memory-intensive operation, and may be very slow on machines with less than 8 GB of RAM.

Once initialized, the Dataset instance contains activation data from nearly 10,000 published neuroimaging articles. But it doesn't yet have any features attached to those data, so let's add some:

	> dataset.add_features('data/features.txt')

Now our Dataset has both activation data and some features we can use to manipulate the data with. In this case, the features are just term-based tags--i.e., words that occur in the abstracts of the articles from which the dataset is drawn (for details, see this [Nature Methods] paper, or the Neurosynth website).

We can now do various kinds of analyses with the data. For example, we can use the features we just added to perform automated large-scale meta-analyses. Let's see what features we have:

	> dataset.get_feature_names()
	['phonetic', 'associative', 'cues', 'visually', ... ]

We can use these features--either in isolation or in combination--to select articles for inclusion in a meta-analysis. For example, suppose we want to run a meta-analysis of emotion studies. We could operationally define a study of emotion as one in which the authors used words starting with 'emo' with high frequency:

	> ids = dataset.get_studies(features='emo*', frequency_threshold=0.001)

Here we're asking for a list of IDs of all studies that use words starting with 'emo' (e.g.,'emotion', 'emotional', 'emotionally', etc.) at a frequency of 1 in 1,000 words or greater (in other words, if an article has 5,000 words of text, it will only be included in our set if it uses words starting with 'emo' at least 5 times).

	> len(ids)
	639

The resulting set includes 639 studies.

Once we've got a set of studies we're happy with, we can run a simple meta-analysis, prefixing all output files with the string 'emotion' to distinguish them from other analyses we might run:

	> from neurosynth.analysis import meta
	> ma = meta.MetaAnalysis(dataset, ids)
	> ma.save_results('some_directory/emotion')

You should now have a set of Nifti-format brain images on your drive that display various meta-analytic results. The image names are somewhat cryptic; see the Documentation for details. It's important to note that the meta-analysis routines currently implemented in Neurosynth aren't very sophisticated; they're designed primarily for efficiency (most analyses should take just a few seconds), and take multiple shortcuts as compared to other packages like ALE or MKDA. But with that caveat in mind (and one that will hopefully be remedied in the near future), Neurosynth gives you a streamlined and quick way of running large-scale meta-analyses of fMRI data.


## Getting help

For a more comprehensive set of examples, see [this tutorial](http://nbviewer.ipython.org/urls/raw.github.com/neurosynth/neurosynth/master/examples/neurosynth_demo.ipynb)--also included in IPython Notebook form in the examples/ folder (along with several other simpler examples).

For bugs or feature requests, please [create a new issue](https://github.com/neurosynth/neurosynth/issues/new). If you run into problems installing or using the software, try posting to the [Neurosynth Google group](https://groups.google.com/forum/#!forum/neurosynthlist) or email [Tal Yarkoni](mailto:tyarkoni@gmail.com).


Changelog.md
## 0.3.7 (July 11, 2018)
Minor release. Changes:
- Update travis-ci config to deploy on tagged build
- Rename output images

## 0.3.6 (July 10, 2018)
This release collects various bug fixes and improvements since the last release. It will probably be the last release (barring occasional maintenance releases) due to deprecation of this package in favor of other alternatives.

## 0.3.5 (March, 8, 2016)
- Reworked base.py: now uses a pandas DataFrame to store activation data, and the Mappable class has been removed.
- Fixed bug in peak-based search
- Minor cleanup and doc improvements

## 0.3.4 (November 8, 2015)
- Updated dependency list and moved dependencies into setup.py
- Added ROI-based decoding method (generalization of voxel-based decoding)
- Added new clustering module with high and low-level APIs
- Python 3 compatibility
- Much more extensive testing via Travis/Coveralls
- Bug fixes/minor improvements:
	- Added thresholding options to Masker
	- Fix broken import statements
	- Removed seaborn dependency

## 0.3.3 (March 9, 2015)
- Added Sphinx docs and readthedocs integration (http://neurosynth.readthedocs.org)
- Improved study selection interface; get_ids_by_*  methods replaced with a single get_studies() method
- Added ability to download data files from within Neurosynth codebase
- Base package now imports key modules into primary namespace (e.g., from Neurosynth import Dataset)
- Expression parsing/study selection now handles N-gram features
- Added deprecation warnings
- Bug fixes/minor improvements:
	- Fixed loss-of-precision error in FDR-thresholded maps
	- Fixed unmasking of 4d volumes
	- Much faster feature count computation
	- Expanded test coverage; now includes real data in tests
	- More consistent output location arguments across codebase
	- Deleted legacy/unused code
	- Fixed naming issue when decoder is fed an array of images

## 0.3.2 (November 7, 2014)
- Dataset, FeatureTable, and ImageTable all now use pandas everywhere
- New datasets in data submodule
- Improved test coverage of Dataset
- Substantially improved Masker object that handles multiple masking layers
- Added functionality to downsample whole brain to isometric grid
- Initial stab at clustering pipeline (work in progress)
- Improved interface for adding new features to Dataset
- Bug fixes/minor improvements:
	- FeatureTable no longer breaks when reading in columns with N-gram names using pandas
	- get_feature_data() now takes reorder argument
	- convert pandas SparseDataFrame to CSR_matrix on data access as workaround for pandas bug
	- Fixed dataset loading/saving bug for datasets loaded/saved using previous versions
	- Fixed the above bug *again*, because it's a persistent little bugger
	- Updated dependency list
	- Fixed a variety of minor bugs caused by minor API changes in new pandas versions

## 0.3.1 (May 15, 2014)
- Data matrices now stored internally as pandas DataFrames
- Began implementation of classifier class (thanks to @zorro4)
- Better/faster data accessor
- Added several usage examples in /examples, including an IPython demo
- Bug fixes/minor tweaks:
	- Random voxel selection routine works properly now
	- All IDs are now strings to avoid type mismatches for numerical IDs
	- Patched README and travis/coveralls definitions to reflect current dependencies

## 0.3.0 (May 26, 2013)

First-ever *official* release following the [Neurosynth hackathon](http://hackathon.neurosynth.org)

- Enabled travis-ci.org and coveralls.io
- Most of the current functionality was developed pre-0.3.0, so, really, just about everything
- Added tests, thanks to @yarikoptic's heroic efforts
- [Testkraut](https://github.com/neurodebian/testkraut) integration

Neurosynth is a work in progress and we reserve the right to occasionally break the API.


examples/neurosynth_demo.py
# -*- coding: utf-8 -*-
# <nbformat>3.0</nbformat>

# <markdowncell>

# # Overview
# 
# In this lab, we'll walk through some of the features of the Neurosynth core tools (http://github.com/neurosynth/neurosynth). By the end, you'll know how to:
# 
# * Import the modules you'll need to perform basic analyses
# * Create a new Neurosynth dataset object from provided text files
# * Run a simple term-based meta-analysis
# * Run a slightly more complicated term-based meta-analysis
# * Perform meta-analytic contrasts
# * Generate seed-based coactivation maps
# * "Decode" your own images
# 
# ## Installation
# 
# We're not going to cover installation here--sorry! See the quickstart guide in the [github repository](http://github.com/neurosynth/neurosynth) for that.
# 
# ## Running this code
# 
# To run the examples below, you have several options:
# 
# * *Run from within IPython Notebook*. This is the preferred approach if you have IPython Notebook installed. Put this file (neurosynth_demo.ipynb) in its own directory. Now launch the IPython dashboard from a command line prompt by typing "ipython notebook" (without quotes), then open the Neurosynth demo. You should now be able to run any cell in the notebook.
# 
# * *Run in IPython*. From a terminal prompt, launch IPython (just type ipython). Now paste the code blocks below.
# 
# * *Write a standalone script*. Save the code blocks below to a separate file and run it as a Python script. This is not recommended as it's non-interactive.
# 
# In all cases, you'll need to make sure that the data files called below (database.txt and features.txt) are accessible at the locations indicated (a subfolder called data/).
# 
# ## Importing modules
# 
# Let's start with the basics. In Python, most modules (i.e., organized chunks of code) are inaccessible by default. Other than a few very basic built-in functions, you'll need to explicitly include every piece of code you want to work with. This may seem cumbersome, but it has the nice effect of (a) making sure you always know exactly what dependencies your code has, and (b) minimizing the memory footprint of your app by only including functionality you know you'll need.
# 
# Like most Python packages, Neurosynth consists of several modules arranged into a semi-sensible tree structure. For this lab, we'll need functionality available in several modules, which we can include like so:

# <codecell>

# Core functionality for managing and accessing data
from neurosynth.base.dataset import Dataset
# Analysis tools for meta-analysis, image decoding, and coactivation analysis
from neurosynth.analysis import meta, decode, network
# The root-level module, included here just so we can set the logging level.
import neurosynth

# We set the logger to display everything at level INFO or above.
neurosynth.set_logging_level('info')

# <markdowncell>

# ## Creating a new dataset
# 
# Next, we create a Dataset, which is the core object most Neurosynth tools operate on. We initialize a Dataset by passing in a database file, which is essentially just a giant list of activation coordinates and associated study IDs. This file can be downloaded from the Neurosynth website or installed from the data submodule (see the Readme for instructions).
# 
# Creating the object will take a few minutes on most machines, as we need to process about 200,000 activations drawn from nearly 6,000 studies. Once that's done, we also need to add some features to the Dataset. Features are just variables associated with the studies in our dataset; literally any dimension a study could be coded on can constitute a feature that Neurosynth can use. In practice, the default set of features included in the data download includes 500 psychological terms (e.g., 'language', 'emotion', 'memory', etc.) that occur with some frequency in the dataset. So when we're talking about the "emotion" feature, we're really talking about how frequently each study in the Dataset uses the word 'emotion' in the full-text of the corresponding article.
# 
# Let's go ahead and create a dataset and add some features:

# <codecell>

# Create a new Dataset instance
dataset = Dataset('data/database.txt')

# Add some features
dataset.add_features('data/features.txt')

# <markdowncell>

# Because this takes a while, we'll save our Dataset object to disk. That way, the next time we want to use it, we won't have to sit through the whole creation operation again:

# <codecell>

dataset.save('dataset.pkl')

# <markdowncell>

# Now in future, instead of waiting, we could just load the dataset from file:

# <codecell>

dataset = Dataset.load('dataset.pkl')   # Note the capital D in the second Dataset--load() is a class method

# <markdowncell>

# ## Doing stuff with Neurosynth
# Now that our Dataset has both activation data and some features, we're ready to start doing some analyses! By design, Neurosynth focuses on facilitating simple, fast, and modestly useful analyses. This means you probably won't break any new ground using Neurosynth, but you should be able to supplement results you've generated using other approaches with a bunch of nifty analyses that take just 2 - 3 lines of code.
# 
# ### Simple feature-based meta-analyses
# The most straightforward thing you can do with Neurosynth is use the features we just loaded above to perform automated large-scale meta-analyses of the literature. Let's see what features we have:

# <codecell>

dataset.get_feature_names()

# <markdowncell>

# If the loading process went smoothly, this should return a list of about 500 terms. We can use these terms--either in isolation or in combination--to select articles for inclusion in a meta-analysis. For example, suppose we want to run a meta-analysis of emotion studies. We could operationally define a study of emotion as one in which the authors used words starting with 'emo' with high frequency:

# <codecell>

ids = dataset.get_ids_by_features('emo*', threshold=0.001)

# <markdowncell>

# Here we're asking for a list of IDs of all studies that use words starting with 'emo' (e.g.,'emotion', 'emotional', 'emotionally', etc.) at a frequency of 1 in 1,000 words or greater (in other words, if an article has 5,000 words of text, it will only be included in our set if it uses words starting with 'emo' at least 5 times). Let's find out how many studies are in our list:

# <codecell>

len(ids)

# <markdowncell>

# The resulting set includes 639 studies.
# 
# Once we've got a set of studies we're happy with, we can run a simple meta-analysis, prefixing all output files with the string 'emotion' to distinguish them from other analyses we might run:

# <codecell>

# Run a meta-analysis on emotion
ids = dataset.get_ids_by_features('emo*', threshold=0.001)
ma = meta.MetaAnalysis(dataset, ids)
ma.save_results('emotion')

# <markdowncell>

# You should now have a set of Nifti-format brain images on your drive that display various meta-analytic results. The image names are somewhat cryptic; see documentation elsewhere for details. It's important to note that the meta-analysis routines currently implemented in Neurosynth aren't very sophisticated; they're designed primarily for efficiency (most analyses should take just a few seconds), and take multiple shortcuts as compared to other packages like ALE or MKDA. But with that caveat in mind (and one that will hopefully be remedied in the near future), Neurosynth gives you a streamlined and quick way of running large-scale meta-analyses of fMRI data. Of course, all of the images you could generate using individual features are already available on the Neurosynth website, so there's probably not much point in doing this kind of thing yourself unless you've defined entirely new features.
# 
# ### More complex feature-based meta-analyses
# 
# Fortunately, we're not constrained to using single features in our meta-analyses. Neurosynth implements a parsing expression grammar, which is a fancy way of saying you can combine terms according to syntactic rules--in this case, basic logical operations.
# 
# For example, suppose we want to restrict our analysis to studies of emotion that do NOT use the terms 'reward' or 'pain', which we might construe as somewhat non-prototypical affective states. Then we could do the following:

# <codecell>

ids = dataset.get_ids_by_expression('emo* &~ (reward* | pain*)', threshold=0.001)
ma = meta.MetaAnalysis(dataset, ids)
ma.save_results('emotion_without_reward_or_pain')
print "Found %d studies." % len(ids)

# <markdowncell>

# This meta-analysis is somewhat more restrictive than the previous one (555 studies instead of 639), and the result should theoretically be at least somewhat more spatially specific.
# 
# There's no inherent restriction on how many terms you combine or how deeply you nest logical expressions within parentheses, but the cardinal of GIGO (garbage in, garbage out) always applies, so if your expression is very specific and the number of studies drops too far (in practice, sensible results are unlikely with fewer than 50 studies), don't expect to see much.
# 
# ### Meta-analytic contrasts
# 
# In addition to various logical operations, one handy thing you can do with Neurosynth is perform meta-analytic contrasts. Meaning, you can identify voxels in which the average likelihood of activation being reported differ for two different sets of studies. For example, let's say you want to meta-analytically contrast studies that use the term 'recollection' with studies that use the term 'recognition'. You can do this by defining both sets of studies separately, and then passing them both to the meta-analysis object:

# <codecell>

# Get the recognition studies and print some info...
recog_ids = dataset.get_ids_by_features('recognition', threshold=0.001)
print "We found %d studies of recognition" % len(recog_ids)

# Repeat for recollection studies
recoll_ids = dataset.get_ids_by_features('recollection', threshold=0.001)
print "We found %d studies of recollection" % len(recoll_ids)

# Run the meta-analysis
ma = meta.MetaAnalysis(dataset, recog_ids, recoll_ids)
ma.save_results('recognition_vs_recollection')

# <markdowncell>

# This produces the same set of maps we've seen before, except the images now represent a meta-analytic contrast between two specific sets of studies, rather than between one set of studies and all other studies in the database.
# 
# It's worth noting that meta-analytic contrasts generated using Neurosynth should be interpreted very cautiously. Remember that this is a meta-analytic contrast rather than a meta-analysis of contrasts. In the above example, we're comparing activation in all studies in which the term recognition shows up often to activation in all studies in which the term recollection shows up often (implicitly excluding studies that use both terms). We are NOT meta-analytically combining direct contrasts of recollection and recognition, which would be a much more sensible thing to do (but is something that can't be readily automated).
# 
# ### Seed-based coactivation maps
# 
# By now you're all familiar with seed-based functional connectivity. We can do something very similar at a meta-analytic level (e.g., Toro et al, 2008, Robinson et al, 2010, Chang et al, 2012) using the Neurosynth data. Specifically, we can define a seed region and then ask what other regions tend to be reported in studies that report activity in our seed region. The Neurosynth tools make this very easy to do. We can either pass in a mask image defining our ROI, or pass in a list of coordinates to use as the centroid of spheres. In this example, we'll do the latter:

# <codecell>

# Seed-based coactivation
network.coactivation(dataset, [[0, 20, 28]], threshold=0.1, outroot='coactivation_from_coords', r=10)

# <markdowncell>

# Here we're generating a coactivation map for a sphere with radius 10 mm centered on an anterior cingulate cortex (ACC) voxel. The threshold argument indicates what proportion of voxels within the ACC sphere have to be activated for a study to be considered 'active'.
# 
# In general, meta-analytic coactivation produces results quite similar--but substantially less spatially specific--than time series-based functional connectivity. Note that if you're only interested in individual points in the brain, you can find precomputed coactivation maps for spheres centered on every gray matter voxel in the brain on the Neurosynth website.
# 
# ### Decoding your own images
# 
# One of the most useful features of Neurosynth is the ability to 'decode' arbitrary images by assessing their similarity to the reverse inference meta-analysis maps generated for different terms. For example, you might wonder whether a group-level z-score map for some experimental contrast is more consistent with recollection or with recognition. You could even use Neurosynth as a simple (but often effective) classifier by running a series of individual subjects through the decoder and picking the class (i.e., term) with the highest similarity. Perhaps the most powerful--though somewhat more computationally intensive--use is to do open-ended decoding. That is, we can take the entire set of features included in the base Neurosynth data download and rank-order them by similarity to each of our input images.
# 
# In this example, we'll decode three insula-based coactivation networks drawn from Chang, Yarkoni, Khaw, & Sanfey (2012). You should substitute your own images into the list below. We assess the similarity of each map with respect to 9 different terms and save the results to a file. Note that if we left the features argument unspecified, the decoder would default to using the entire set of 500+ features (which will take a few minutes on most machines unless you've pregenerated the feature maps--but that's for a different tutorial).

# <codecell>

# Decode images
decoder = decode.Decoder(dataset, features=['taste', 'disgust', 'emotion', 'auditory', 'pain', 'somatosensory', 'conflict', 'switching', 'inhibition'])
data = decoder.decode(['pIns.nii.gz', 'vIns.nii.gz', 'dIns.nii.gz'], save='decoding_results.txt')

# <markdowncell>

# In decoding_results.txt, we have features in rows, and input images in columns. By default, each cell reflects the pearson correlation between the corresponding input image (i.e., the column) and reverse inference meta-analysis map (i.e., the row). Sort the columns in descending order and you've got a crude but potentially quite useful open-ended decoding of your images. Mind you, if you're lazy, you can achieve the same thing by uploading your images to [NeuroVault](http://neurovault.org) and then using the (currently experimental) [decode](http://neurosynth.org) function on the [Neurosynth website](http://neurosynth.org).

# <codecell>


